# **MÓDULO 20 - Projeto de Credit Score - Naive Bayes**


No módulo 17 vocês realizaram a primeira etapa do projeto de crédito de vocês.
Então fizeram o tratamendo dos dados, balancearam as classes, transformaram as variáveis categóricas e separam base de treino e teste.
Nessa aula aplicaremos o algoritmo de naive bayes a base de vocês afim de tentarmos trazer previsões do score de crédito.

**IMPORTANTE:** Não se esqueçam de ao enviar o código de vocês para os tutores, enviarem as bases, pois como cada um de vocês realizou as alterações de tratamento indidualmente o tutor precisa ter acesso aos seus dados individuais.

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.metrics import confusion_matrix
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score

Durante a aula nossa variável a ser prevista (churn) continha apenas 2 categorias, a base de vocês contém mais. O Naive Bayes pode ser aplicado para problemas de classificação com múltiplas classes da mesma forma que para problemas de classificação binária. O Naive Bayes é um algoritmo de classificação probabilístico que calcula a probabilidade de uma amostra pertencer a cada classe e seleciona a classe com a maior probabilidade como a previsão final.
Em resumo, o Naive Bayes pode ser aplicado da mesma maneira para problemas de classificação com múltiplas classes, e os mesmos princípios se aplicam em termos de treinamento, avaliação e aplicação do modelo.

## Carregamento da Base Tratada 

In [2]:
df = pd.read_csv("C:/Users/karin/OneDrive/Área de Trabalho/EBAC/CREDIT_SCORE_PROJETO_PARTE1_preparada.csv")

In [3]:
df.head(10)

Idade     Renda  Escolaridade  Qtd_Filhos  Genero_Masculino  \
0   45.0  115000.0             2           3             False   
1   36.0  135000.0             2           0              True   
2   45.0  110000.0             2           3             False   
3   36.0   47500.0             4           0              True   
4   26.0   40000.0             4           0             False   
5   44.0   75000.0             3           0              True   
6   42.0  105000.0             3           0              True   
7   31.0   65000.0             2           0              True   
8   41.0  110000.0             0           0              True   
9   27.0   37500.0             1           0             False   

   Estado_Civil_Solteiro(a)  Residencia_Própria  Score_Credito  
0                     False                True              0  
1                     False                True              0  
2                     False                True              0  
3                      True               False              2  
4                      True               False              2  
5                      True                True              0  
6                      True                True              0  
7                      True               False              2  
8                      True                True              0  
9                      True               False              1

In [4]:
df.columns


Index(['Idade', 'Renda', 'Escolaridade', 'Qtd_Filhos', 'Genero_Masculino',
       'Estado_Civil_Solteiro(a)', 'Residencia_Própria', 'Score_Credito'],
      dtype='object')

# 1) Comece carregando as bases de treino (X e y) e teste (X e y).
Verifique se o número de linhas condiz, se as variáveis estão corretas sendo apenas a de score para y e as demais nas bases de X e por último, se Y está balanceada no teste.

### 1) Separar X e y
variável-alvo(target) será : Score_Credito

In [6]:
# Separar variáveis preditoras e a variável alvo
X = df.drop('Score_Credito', axis=1)
y = df['Score_Credito']


# 2) Aplique o algoritmo de Naive Bayes aos dados de treinamento.

### Aqui dividirei os dados em treino e teste e treinamos o modelo Gaussian Naive Bayes com os dados de treino.


In [7]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

# Separar em treino e teste
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

# Instanciar e treinar o modelo
modelo_nb = GaussianNB()
modelo_nb.fit(X_train, y_train)


GaussianNB()

# 3) Faça a avaliação do modelo com os dados de treinamento.
Traga a acurácia, recall e plote a matriz de confusão. Não se esqueça de avaliar com suas palavras o desempenho do modelo, interpretando as métricas.

### Avaliarei o modelo com o conjunto de treino utilizando métricas como acurácia, recall e matriz de confusão.

Dica: Para calcularmos o recall em classificação multi classe precisamos usar o atributo macro:
recall = recall_score(y_train, y_pred_train, average='macro')

In [8]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Previsão com base de treino
y_pred_train = modelo_nb.predict(X_train)

# Avaliação
print("Matriz de Confusão (Treino):")
print(confusion_matrix(y_train, y_pred_train))

print("\nRelatório de Classificação (Treino):")
print(classification_report(y_train, y_pred_train))

print(f"Acurácia (Treino): {accuracy_score(y_train, y_pred_train):.2f}")#seu código aqui

Matriz de Confusão (Treino):
[[56  0  1]
 [ 0 57  0]
 [ 0  2 54]]

Relatório de Classificação (Treino):
              precision    recall  f1-score   support

           0       1.00      0.98      0.99        57
           1       0.97      1.00      0.98        57
           2       0.98      0.96      0.97        56

    accuracy                           0.98       170
   macro avg       0.98      0.98      0.98       170
weighted avg       0.98      0.98      0.98       170

Acurácia (Treino): 0.98


### Interpretação 
O modelo acertou 98% das amostras de treino.

As classes estão bem balanceadas e o modelo teve bom desempenho em todas elas.

O recall da classe 1 foi 100%, ou seja, todos os exemplos dessa classe foram corretamente reconhecidos.

Pouquíssimos erros (confusões) entre as classes.

# 4) Aplique o modelo aos dados de teste e realize a avaliação dos resultados, da mesma forma que fez acima. Não se esqueça de avaliar com as suas palavras e comparar o desempenho da base treino com a teste.

### Agora vou aplicar o modelo treinado aos dados de teste.  
Irei comparar os resultados com os obtidos no treino, observando possíveis diferenças entre as métricas (como overfitting ou perda de desempenho).  
Analisarei novamente a matriz de confusão, precisão, recall e f1-score.

In [9]:
# Previsão no conjunto de teste
y_pred_test = modelo_nb.predict(X_test)

# Avaliação
print("Matriz de Confusão (Teste):")
print(confusion_matrix(y_test, y_pred_test))

print("\nRelatório de Classificação (Teste):")
print(classification_report(y_test, y_pred_test))

print(f"Acurácia (Teste): {accuracy_score(y_test, y_pred_test):.2f}")


Matriz de Confusão (Teste):
[[24  0  0]
 [ 0 24  0]
 [ 0  2 23]]

Relatório de Classificação (Teste):
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        24
           1       0.92      1.00      0.96        24
           2       1.00      0.92      0.96        25

    accuracy                           0.97        73
   macro avg       0.97      0.97      0.97        73
weighted avg       0.97      0.97      0.97        73

Acurácia (Teste): 0.97


# 5) Descreva com suas palavras o projeto desenvolvido nessa atividade e qual o nosso objetivo principal ao aplicarmos o algoritmo de naive bayes a base de crédito.
Utilize pelo menos 4 linhas.

Dica: Caso você ainda esteja tendo dificuldade em visualizar a aplicação dos projetos e objetivo, consulte seus tutores!

## 5. Conclusão do Projeto

Ao longo deste projeto, apliquei o algoritmo Naive Bayes a uma base de dados de score de crédito previamente tratada.  
Meu objetivo foi prever corretamente a categoria de score de novos clientes com base em suas características pessoais e financeiras.  
O modelo obteve desempenho muito bom, com acurácia de 98% no treino e 97% no teste, o que indica boa generalização.  
Além disso, todas as classes foram previstas corretamente com alta precisão e recall, demonstrando que o algoritmo é eficiente mesmo com a suposição de independência entre variáveis.  
Este projeto mostrou que o Naive Bayes é uma ferramenta rápida, interpretável e útil para problemas de classificação com dados estruturados.
